core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/CachingChunkManager.java

```diff
@@ -0,0 +1,125 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.chunkmanager;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+
+import io.aiven.kafka.tieredstorage.manifest.SegmentManifest;
+import io.aiven.kafka.tieredstorage.storage.StorageBackendException;
+
+import com.github.benmanes.caffeine.cache.AsyncCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.github.benmanes.caffeine.cache.Scheduler;
+import com.github.benmanes.caffeine.cache.Weigher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class CachingChunkManager<T> implements Configurable, ChunkManager {
+    private static final Logger log = LoggerFactory.getLogger(CachingChunkManager.class);
+    private static final long GET_TIMEOUT_SEC = 10;
+    private final ChunkManager chunkManager;
+    private final Executor executor;
+    protected AsyncCache<ChunkKey, T> cache;
+
+    public CachingChunkManager(final ChunkManager chunkManager, final Executor executor) {
+        this.chunkManager = chunkManager;
+        this.executor = executor;
+    }
+
+    public InputStream getChunk(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                final SegmentManifest manifest,
+                                final int chunkId) throws StorageBackendException, IOException {
+        final Uuid id = remoteLogSegmentMetadata.remoteLogSegmentId().id();
+        final ChunkKey chunkKey = new ChunkKey(id, chunkId);
+        final AtomicReference<InputStream> result = new AtomicReference<>();
+        try {
+            return cache.asMap().compute(chunkKey, (key, val) -> CompletableFuture.supplyAsync(() -> {
+                if (val == null) {
+                    try {
+                        final InputStream chunk = chunkManager.getChunk(remoteLogSegmentMetadata, manifest, chunkId);
+                        final T t = this.cacheChunk(chunkKey, chunk);
+                        result.getAndSet(chunkToStream(t));
+                        return t;
```

Comment:
We may save one transformation when writing to cache -- unless streams are different?:
```suggestion
                        final InputStream chunk = chunkManager.getChunk(remoteLogSegmentMetadata, manifest, chunkId);
                        cacheChunk(chunkKey, chunk);
                        result.getAndSet(chunk);
                        return t;
```

Reply:
We can not, when we are caching the stream is read.